File: lib/download.py
"""Download LWJGL/JOML jars into the workspace lib folder."""

import urllib.request
import urllib.error
from urllib.parse import urlsplit
import os

LWJGL_VERSION = "3.3.6"
MAVEN_BASE = "https://repo1.maven.org/maven2"
LIB_DIR = os.path.dirname(os.path.abspath(__file__))

# Ensure downloads go to the same folder used by run_java.bat classpath: lib/*
os.chdir(LIB_DIR)

# Remove old LWJGL jars first so Java doesn't mix versions on classpath
for name in os.listdir(LIB_DIR):
    if name.startswith("lwjgl-") and name.endswith(".jar"):
        try:
            os.remove(os.path.join(LIB_DIR, name))
        except OSError:
            pass


def lwjgl_url(module: str, artifact: str) -> str:
    """Build a Maven Central URL for a LWJGL artifact."""
    return (
        f"{MAVEN_BASE}/org/lwjgl/{module}/{LWJGL_VERSION}/"
        f"{artifact}-{LWJGL_VERSION}.jar"
    )


urls = [
    lwjgl_url("lwjgl", "lwjgl"),
    lwjgl_url("lwjgl", "lwjgl-natives-windows"),
    lwjgl_url("lwjgl-glfw", "lwjgl-glfw"),
    lwjgl_url("lwjgl-glfw", "lwjgl-glfw-natives-windows"),
    lwjgl_url("lwjgl-opengl", "lwjgl-opengl"),
    lwjgl_url("lwjgl-opengl", "lwjgl-opengl-natives-windows"),
    f"{MAVEN_BASE}/org/joml/joml/1.10.5/joml-1.10.5.jar",
    lwjgl_url("lwjgl-stb", "lwjgl-stb"),
    lwjgl_url("lwjgl-stb", "lwjgl-stb-natives-windows"),
]

for url in urls:
    filename = os.path.basename(urlsplit(url).path)
    filepath = os.path.join(LIB_DIR, filename)
    print(f"Downloading {filename}...")
    try:
        urllib.request.urlretrieve(url, filepath)
    except (urllib.error.URLError, OSError) as e:
        print(f"Failed to download {filename}: {e}")

print("All downloads complete!")

File: lib/download.js
const fs = require('fs');
const https = require('https');
const path = require('path');

const javalib = path.join(__dirname, 'javalib');
if (!fs.existsSync(javalib)) {
    fs.mkdirSync(javalib);
}

const urls = [
    "https://repo1.maven.org/maven2/org/lwjgl/lwjgl/3.3.3/lwjgl-3.3.3.jar",
    "https://repo1.maven.org/maven2/org/lwjgl/lwjgl/3.3.3/lwjgl-3.3.3-natives-windows.jar",
    "https://repo1.maven.org/maven2/org/lwjgl/lwjgl-glfw/3.3.3/lwjgl-glfw-3.3.3.jar",
    "https://repo1.maven.org/maven2/org/lwjgl/lwjgl-glfw/3.3.3/lwjgl-glfw-3.3.3-natives-windows.jar",
    "https://repo1.maven.org/maven2/org/lwjgl/lwjgl-opengl/3.3.3/lwjgl-opengl-3.3.3.jar",
    "https://repo1.maven.org/maven2/org/lwjgl/lwjgl-opengl/3.3.3/lwjgl-opengl-3.3.3-natives-windows.jar",
    "https://repo1.maven.org/maven2/org/joml/joml/1.10.5/joml-1.10.5.jar",
    "https://repo1.maven.org/maven2/org/lwjgl/lwjgl-stb/3.3.3/lwjgl-stb-3.3.3.jar",
    "https://repo1.maven.org/maven2/org/lwjgl/lwjgl-stb/3.3.3/lwjgl-stb-3.3.3-natives-windows.jar"
];

async function download(url, dest) {
    return new Promise((resolve, reject) => {
        const file = fs.createWriteStream(dest);
        https.get(url, (response) => {
            if (response.statusCode >= 300 && response.statusCode < 400 && response.headers.location) {
                return download(response.headers.location, dest).then(resolve).catch(reject);
            }
            response.pipe(file);
            file.on('finish', () => {
                file.close(resolve);
            });
        }).on('error', (err) => {
            fs.unlink(dest, () => reject(err));
        });
    });
}

(async () => {
    for (const url of urls) {
        const filename = path.basename(url);
        const dest = path.join(javalib, filename);
        console.log(`Downloading ${filename}...`);
        try {
            await download(url, dest);
        } catch (e) {
            console.error(e);
        }
    }
    console.log("Done!");
})();
